client/src/modules/display_metadata/display_metadata.js

```diff
@@ -0,0 +1,208 @@
+    .catch(Notify.handleError);
+
+  if ($state.params.id && $state.params.patient) {
+    // Prevent a non-patient form from being used to collect non-patient data
+    vm.patientData = true;
+  } else {
+    vm.patientData = false;
+
+    if (vm.collectorId) {
+      loadGrid();
+    }
+  }
+
+  // options for the UI grid
+  vm.gridOptions = {
+    appScopeProvider  : vm,
+    enableColumnMenus : false,
+    fastWatch         : true,
+    flatEntityAccess  : true,
+    enableSorting     : true,
+    showColumnFooter  : true,
+    onRegisterApi     : onRegisterApiFn,
+    enableHorizontalScrollbar : uiGridConstants.scrollbars.WHEN_NEEDED,
+  };
+
+  function onRegisterApiFn(gridApi) {
+    vm.gridApi = gridApi;
+  }
+
+  function toggleFilter() {
+    vm.filterEnabled = !vm.filterEnabled;
+    vm.gridOptions.enableFiltering = vm.filterEnabled;
+    vm.gridApi.core.notifyDataChange(uiGridConstants.dataChange.COLUMN);
+  }
+
+  // bind methods
+  vm.edit = edit;
+  vm.remove = remove;
+
+  function edit(data) {
+    if (!$state.params.patient) {
+      $state.go('display_metadata.edit', { id : data.data_collector_management_id, uuid : data.uuid });
+    } else if ($state.params.patient) {
+      $state.go('display_metadata.patientEdit', {
+        id : data.data_collector_management_id,
+        uuid : data.uuid,
+        patient : $state.params.patient,
+      });
+    }
+  }
+
+  function remove(uuid) {
+    ModalService.confirm('FORM.DIALOGS.CONFIRM_DELETE')
+      .then(() => {
+        DisplayMetadata.delete(uuid)
+          .then(() => {
+            Notify.success('FORM.INFO.DELETE_SUCCESS');
+            loadGrid();
+          })
+          .catch(Notify.handleError);
+      });
+  }
+
+  function handleError(error) {
+    vm.hasError = true;
+    Notify.handleError(error);
+  }
+
+  // load user grid
+  function loadGrid() {
+    toggleLoadingIndicator();
+    // Reuse cash unless patient data is consulted
+    if (cache.collector && !vm.collectorId && !$state.params.patient) {
+      vm.collectorId = cache.collector.id;
+    }
+    const changesLength = Object.keys(vm.changes).length;
+
+    if (cache.changes && !changesLength) {
+      vm.changes = cache.changes;
+    }
+
+    if (!$state.params.id && !vm.collectorId) {
+      search();
+    } else if (vm.collectorId && !$state.params.id) {
+      vm.hasError = false;
+      vm.loading = true;
+
+      SurveyForm.read(null, { data_collector_management_id : vm.collectorId })
+        .then((survey) => {
+          vm.filterElements = DisplayMetadata.displayFilters(survey, vm.changes);
+
+          vm.params = {
+            data_collector_management_id : vm.collectorId,
+            changes : vm.changes,
+          };
+
+          return DisplayMetadata.read(null, vm.params);
+        })
+        .then((data) => {
+          vm.gridOptions.columnDefs = data.columns;
+          vm.gridOptions.data = data.surveyData;
+
+          return DataCollectorManagement.read(vm.collectorId);
```

Comment:
You have a nice promise chain here that sends a request for:
 1. `SurveyForm.read()`
 2. `DisplayMetadata.read()`
 3. `DataCollectorManagement.read()`

However, it doesn't look like any of these actually depend on one another.  So, let's send them in parallel using `Promise.all()`!

```js
vm.params = {/* etc  */};
return Promise.all([
  SurveyForm.read(null, { data_collector_management_id : vm.collectorId }),
  DisplayMetadata.read(null, vm.params),
  DataCollectorManagement.read(vm.collectorId),
])
.then(([survey, data, collector]) => {
  // process all queries...
});
```

Reply:
@jniles, i try this options but the displaying became too slow